src/librustc_trans/trans/base.rs

```diff
@@ -848,7 +918,16 @@ pub fn from_arg_ty(bcx: Block, val: ValueRef, ty: Ty) -> ValueRef {
 
 pub fn to_arg_ty(bcx: Block, val: ValueRef, ty: Ty) -> ValueRef {
     if ty.is_bool() {
-        Trunc(bcx, val, Type::i1(bcx.ccx()))
+        let val = Value(val);
+        if let Some(zext) = val.as_zext_inst() {
+            if let Some(val) = zext.get_operand(0) {
+                let valty = val_ty(val.get());
+                if valty == Type::i1(bcx.ccx()) {
+                    return val.get();
```

Comment:
The primary reason for this change is so the likely/unlikely intrinsics actually work. LLVM is very particular about it and requires that the result from the intrinsic go straight into the condition for the branch. The changes to function calls handles straight `if likely(cond)`, but no `if unsafe { likely(cond) }`. This, combined with the load elision, means that first the load from the temporary is skipped, then the truncation of `zext` bool is skipped.